### PR TITLE
added comparison to ibm watson

### PR DESCRIPTION
**Pull request by feliciatrinh**

### CODE REVIEW

File: benchmark/engine.py [hunk 1]
```diff
@@ -77,11 +83,13 @@ def __str__(self):
         raise NotImplementedError()
 
     @classmethod
-    def create(cls, engine_type, project_id):
+    def create(cls, engine_type, gcp_project_id, ibm_model_id, ibm_custom_id):
```
Comment (kenarsa) [inline benchmark/engine.py]: let's use keyword arguments here. 

def create(cls, engine_type, **kwargs) ...

then extract what you need for each engine.

File: benchmark/custom.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
+import json
```
Comment (kenarsa) [inline benchmark/custom.py]: let's make all this part of the engine code.

File: benchmark/benchmark.py [hunk 3]
```diff
@@ -21,14 +21,23 @@ def _path(x):
...
     if args.gcp_credential_path is not None:
         os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = args.gcp_credential_path
 
     if args.ibm_credential_path is not None:
-        os.environ['IBM_CREDENTIALS_FILE'] = args.ibm_credential_path
-
-    engine = NLUEngine.create(NLUEngines[args.engine_type], args.gcp_project_id, args.ibm_model_id, args.ibm_custom_id)
+        with open(args.ibm_credential_path) as f:
+            for line in f:
+                if not line.strip():
```
Comment (kenarsa) [inline benchmark/benchmark.py]: are you checking if the line is empty? If yes, please do it explicitly:

if len(line.strip()) > 0:
...

File: benchmark/benchmark.py [hunk 4]
```diff
@@ -21,14 +21,23 @@ def _path(x):
...
+                    continue
+                key, value = line.strip().split('=', 1)
```
Comment (kenarsa) [inline benchmark/benchmark.py]: .split('=', maxsplit=1)

for literal parameters always use the argument name

File: benchmark/engine.py [hunk 5]
```diff
@@ -11,8 +11,10 @@
 from ibm_watson import NaturalLanguageUnderstandingV1
 from ibm_watson.natural_language_understanding_v1 import Features, EntitiesOptions
 import soundfile
+import requests
```
Comment (kenarsa) [inline benchmark/engine.py:10]: is this PEP compliant? (the order of imports)

File: benchmark/engine.py [hunk 6]
```diff
@@ -83,13 +85,20 @@ def __str__(self):
         raise NotImplementedError()
 
     @classmethod
-    def create(cls, engine_type, gcp_project_id, ibm_model_id, ibm_custom_id):
+    def create(cls, **kwargs):
```
Comment (kenarsa) [inline benchmark/engine.py]: you probably want it to be:

def create(engine_type, **kwargs):
...

engine_type is a mandatory parameter

File: benchmark/engine.py [hunk 7]
```diff
@@ -83,13 +85,20 @@ def __str__(self):
...
 
     @classmethod
-    def create(cls, engine_type, gcp_project_id, ibm_model_id, ibm_custom_id):
+    def create(cls, **kwargs):
+        engine_type = NLUEngines[kwargs['engine_type']]
+        gcp_project_id = kwargs['gcp_project_id']
+        ibm_model_id = kwargs['ibm_model_id']
+        ibm_custom_id = kwargs['ibm_custom_id']
+        apikey = kwargs['apikey']
+        stt_url = kwargs['stt_url']
```
Comment (kenarsa) [inline benchmark/engine.py]: you should look for keys only when constructing the engine of the specific type. what happens if you are constructing Lex and none of these are in the dictionary?

File: benchmark/engine.py [hunk 8]
```diff
@@ -185,13 +194,96 @@ def __str__(self):
...
+            response = requests.get(uri, auth=(self._username, self._apikey), verify=False, headers=self._headers)
+            status = response.json()['status']
+            time_to_run += 10
+
+        if status != 'available':
+            raise RuntimeError()
+
+        print("Training custom model complete")
+
+    def train_language_model(self):
```
Comment (kenarsa) [inline benchmark/engine.py]: this should be a protected method as its only called by constructor

File: benchmark/engine.py [hunk 9]
```diff
@@ -185,13 +194,96 @@ def __str__(self):
...
+
+    def train(self):
+        uri = self._stt_url + "/v1/customizations/" + self._custom_id + "/train"
+        response = requests.post(uri, auth=(self._username, self._apikey),
+                                 verify=False, data=json.dumps({}).encode('utf-8'))
+
+        if response.status_code != 200:
+            raise RuntimeError("Failed to start training custom model")
+
+    def get_training_status(self):
```
Comment (kenarsa) [inline benchmark/engine.py]: this should be a protected method as its only called by constructor

File: benchmark/engine.py [hunk 10]
```diff
@@ -185,13 +194,96 @@ def __str__(self):
...
+                                     headers=self._headers, data=f)
+
+        if response.status_code != 201:
+            print(response.text)
+            raise RuntimeError("Failed to add corpus file")
+
+        print("Added corpus file")
+        return uri
+
+    def get_corpus_status(self, uri):
```
Comment (kenarsa) [inline benchmark/engine.py]: this should be a protected method as its only called by constructor

File: benchmark/engine.py [hunk 11]
```diff
@@ -185,13 +194,96 @@ def __str__(self):
...
+
+        if response.status_code != 201:
+            print(response.text)
+            raise RuntimeError("Failed to create model")
+
+        custom_id = response.json()['customization_id']
+        print("Model customization id: ", custom_id)
+        return custom_id
+
+    def add_corpus(self):
```
Comment (kenarsa) [inline benchmark/engine.py]: this should be a protected method as its only called by constructor

File: benchmark/engine.py [hunk 12]
```diff
@@ -185,13 +194,96 @@ def __str__(self):
...
+        self._stt_url = stt_url
+        self._headers = {'Content-Type': "application/json"}
         if custom_id is None:
-            self._custom_id = create_language_model()
+            self._custom_id = self.create_language_model()
+            self.train_language_model()
         else:
             self._custom_id = custom_id
 
+    def create_language_model(self):
```
Comment (kenarsa) [inline benchmark/engine.py]: this should be a protected method as its only called by constructor

**Review by kenarsa — CHANGES_REQUESTED** [review]

File: benchmark/engine.py (hunk 6 above)
Comment (feliciatrinh) [reply benchmark/engine.py]: My kwargs dictionary includes all the command line arguments. I get an error "create() got multiple values for argument 'engine_type' " if I do create(engine_type, **kwargs). And I enforce that engine_type is a required argument in benchmark.py, so engine_type will always have a non-null value in kwargs.

File: benchmark/engine.py (hunk 7 above)
Comment (feliciatrinh) [reply benchmark/engine.py]: You're right. The middle 3 variables would have None values, but I would definitely get a KeyError with the last two.

File: benchmark/engine.py (hunk 6 above)
Comment (kenarsa) [reply benchmark/engine.py]: ah got it! that makes sense. But do we need to change the signature of .create? I mean can we do:

from main:

.create(**args)

and keep the signature in engine.py as

.create(engine_type, **kwargs)

if not. don't worry and just ignore this comment

**Pull request merged**